### PR TITLE
Fixed issue where CosmosDiagnosticsContext is null when diagnostics are sampled out.

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description

When diagnostics are sampled out right now not even the CosmosDiagnosticsContext gets published. Instead sampling should only effect emitting diagnostics (calling DiagnosticsHandler) - but still allow apps to access the CosmosDiagnosticsContext.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosDiagnosticsE2ETest.java [hunk 1]
```diff
@@ -177,10 +204,46 @@ public void onlyCustomLoggerWithCustomConfig() {
...
+            .clientTelemetryConfig(
+                new CosmosClientTelemetryConfig()
+                    .diagnosticsThresholds(
+                        new CosmosDiagnosticsThresholds()
+                            .setPointOperationLatencyThreshold(Duration.ofMillis(100))
+                            .setNonPointOperationLatencyThreshold(Duration.ofMillis(2000))
+                            .setRequestChargeThreshold(100)
+                    )
+                    .diagnosticsHandler(capturingLogger)
+                    .sampleDiagnostics(0)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosDiagnosticsE2ETest.java:238]: Can we also add a test case where we actually sampleDiagnostics with value > 0 and < 1?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/CosmosPagedFlux.java [hunk 2]
```diff
@@ -150,9 +150,7 @@ private Flux<FeedResponse<T>>  wrapWithTracingIfEnabled(CosmosPagedFluxOptions p
         FeedOperationState state = pagedFluxOptions.getFeedOperationState();
         DiagnosticsProvider tracerProvider = state != null ? state.getDiagnosticsProvider() : null;
         Object lockHolder = new Object();
-        if (tracerProvider == null ||
-            !tracerProvider.isEnabled()
-            || tracerProvider.shouldSampleOutOperation(pagedFluxOptions)) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/CosmosPagedFlux.java:155]: why are we removing it from here?

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosDiagnosticsE2ETest.java (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosDiagnosticsE2ETest.java:238]: what purpose would that have? That test would be flaky by design - unless you use a high enough iteration count to force hitting each of the two code paths being hit?

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @FabianMeiswinkel

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/CosmosPagedFlux.java (hunk 2 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/CosmosPagedFlux.java:155]: nit: If we are removing this then the `shouldSampleOutOperation` overload with `CosmosPageFluxOptions` remains unused and can be removed.  Also, a CosmosDiagnosticsContext can be constructed independently of tracer being enabled or not? Just curious on the coupling between the two concepts.

**Review by jeet1995 — APPROVED** [review]
LGTM - thanks!

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosDiagnosticsE2ETest.java (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/CosmosDiagnosticsE2ETest.java:238]: Discussed offline - validation has been added for sampling rate 1 and 0 - so all code paths are covered by the tests added.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/CosmosPagedFlux.java (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/CosmosPagedFlux.java:155]: If there is any diagnostics handler provided than the tracer is not disabled. When no tracing or no diagnostics are enabled then disabling the additional reactor overhead was intentional - and I would keep it for now (the overhead is >5 percent after all).

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/CosmosPagedFlux.java:155]: So, yes - in theory you are right - but it is not possible without adding some overhead right now - and for the hotfix I think we should limit the scope as much as reasonable.

**Review by jeet1995 — APPROVED** [review]
LGTM on the latest changes - thanks!

**Comment by FabianMeiswinkel** [comment]
Change feed split tests failing due to known flakiness

**Pull request merged**